src/main/java/org/opensearch/security/privileges/PitAccessEvaluator.java

```diff
@@ -0,0 +1,344 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.privileges;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableSet;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.ActionRequest;
+import org.opensearch.action.LatchedActionListener;
+import org.opensearch.action.admin.indices.segments.PitSegmentsRequest;
+import org.opensearch.action.search.DeletePitRequest;
+import org.opensearch.action.search.GetAllPitNodesRequest;
+import org.opensearch.action.search.GetAllPitNodesResponse;
+import org.opensearch.action.search.ListPitInfo;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.security.OpenSearchSecurityPlugin;
+import org.opensearch.security.resolver.IndexResolverReplacer;
+import org.opensearch.security.securityconf.SecurityRoles;
+import org.opensearch.security.user.User;
+
+/**
+ * This class evaluates privileges for point in time (Delete and List all) operations
+ */
+public class PitAccessEvaluator {
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+    private boolean isDebugEnabled = log.isDebugEnabled();
+
+    public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final ClusterService clusterService,
+                                                final User user, final SecurityRoles securityRoles, final String action,
+                                                final IndexNameExpressionResolver resolver,
+                                                final PrivilegesEvaluatorResponse presponse) {
+
+        // Skip pit evaluation for "NodesGetAllPITs" action, since it fetches all PITs across the cluster
+        // for privilege evaluation
+        if(action.startsWith("cluster:admin")) {
+            return presponse;
```

Comment:
How is the permission for `NodesGetAllPITs` evaluated if requests bypass privilege evaluator? Or is there no need to have access control on `NodesGetAllPITs`?

Reply:
It'll still get evaluated, we are just not evaluating it as part of pit access evaluator.

We're not marking presponse as complete, so it'll continue down the privilegeEvaluator code flow.

In short, User needs the action permission to execute nodes get all pits.

Reply:
@bharath-techie looks like the permission for `NodesGetAllPITs` is evaluated in [PitPrivilegesEvaluator.java](https://github.com/opensearch-project/security/pull/2053/files#diff-eb0f476c4343f7d278cdcc617b1de506134909cdccb93baf4effd7c914d0b043)? Would you please point me to where it gets evaluated and explain a bit more about how it gets evaluated?

Reply:
@bharath-techie looks like the permission for `NodesGetAllPITs` is designed to be evaluated in [PitPrivilegesEvaluator.java](https://github.com/opensearch-project/security/pull/2053/files#diff-eb0f476c4343f7d278cdcc617b1de506134909cdccb93baf4effd7c914d0b043)? Would you please point me to where it gets evaluated and explain a bit more about how it gets evaluated?

Reply:
As long as we don't mark the presponse as complete, it'll continue down the flow in `PrivilegesEvaluator`

```
 // check access for point in time requests
        if(pitAccessEvaluator.evaluate(request, clusterService, user, securityRoles,
                action0, resolver, dcm.isDnfofForEmptyResultsEnabled(), presponse).isComplete()) {
            return presponse;
        }
```

It'll go inside this if block for evaluation after we return without marking it complete https://github.com/opensearch-project/security/pull/2053/files#diff-e746c76b234a31b6f95d428b7c40dabf805b609b110c560902c967db8f83fa7aR303